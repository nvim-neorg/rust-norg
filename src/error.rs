use chumsky::error::Simple;

use crate::{stage_1::NorgToken, stage_2::NorgBlock, NorgASTFlat};

/// Represents errors that can occur during the parsing process across different stages.
#[derive(Debug)]
pub enum NorgParseError {
    Stage1(Vec<Simple<char>>),
    Stage2(Vec<Simple<NorgToken>>),
    Stage3(Vec<Simple<NorgBlock>>),
    Stage4(Vec<Simple<NorgASTFlat>>),
}

impl From<Vec<Simple<char>>> for NorgParseError {
    fn from(error: Vec<Simple<char>>) -> Self {
        NorgParseError::Stage1(error)
    }
}

impl From<Vec<Simple<NorgToken>>> for NorgParseError {
    fn from(error: Vec<Simple<NorgToken>>) -> Self {
        NorgParseError::Stage2(error)
    }
}

impl From<Vec<Simple<NorgBlock>>> for NorgParseError {
    fn from(error: Vec<Simple<NorgBlock>>) -> Self {
        NorgParseError::Stage3(error)
    }
}

impl From<Vec<Simple<NorgASTFlat>>> for NorgParseError {
    fn from(error: Vec<Simple<NorgASTFlat>>) -> Self {
        NorgParseError::Stage4(error)
    }
}
