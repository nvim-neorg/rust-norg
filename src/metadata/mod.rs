use chumsky::Parser;
pub use stage_1::NorgMeta;

use crate::error::NorgParseError;

pub mod stage_1;

/// Parses the given input string to produce an AST for the metadata
pub fn parse_metadata(input: &str) -> Result<NorgMeta, NorgParseError> {
    // don't ask me why this fixes it. I don't even care
    let processed = input.replace("\n]", "\n ]");
    let processed = format!("{{\n{}\n}}\n", processed.trim());
    Ok(stage_1::meta_parser().parse(processed)?)
}

#[cfg(test)]
mod tests {
    use insta::assert_yaml_snapshot;
    use itertools::Itertools;

    use crate::metadata::parse_metadata;

    #[test]
    fn common_metadata() {
        let examples: Vec<_> = [
            "
            title: Sunday November 17, 2024
            description: We Cooked
            authors: benlubas
            categories: journal
            created: 2024-11-18
            updated: 2024-11-18T17:58:21-0500
            version: 1.1.1
            ",
            "
            title: Neorg Extras
            description: Extra lua code to configure Neorg
            authors: benlubas
            categories: [
              neorg
              nvim
              config
            ]
            tangle: {
              languages: {
                lua: ~/github/.dotfiles/nvim/lua/benlubas/neorg/extras.lua
              }
              delimiter: heading
            }
            created: 2024-05-03T13:36:42-0500
            updated: 2024-10-27T11:12:32-0500
            version: 1.1.1
            ",
        ]
        .into_iter()
        .map(|example| example.to_string() + "\n")
        .map(|str| parse_metadata(&str))
        .try_collect()
        .unwrap();

        assert_yaml_snapshot!(examples);
    }

    #[test]
    fn arrays() {
        let examples: Vec<_> = [
            "empty_arr: []
             arr: [

             ]",
            "
            categories: [
              one
              two
              45
            ]",
            "
            arr: [
              arrays can contain everything
              5
              -5
              6.02e27
              nil
              {
                x: y
                a: [
                  b
                ]
              }
              []
              [
                hi
                hi
              ]
            ]",
            "arr:[]\na2:[\n]\na3:[\nhi\n]\nx: y",
        ]
        .into_iter()
        .map(|example| example.to_string() + "\n")
        .map(|str| parse_metadata(&str))
        .try_collect()
        .unwrap();

        assert_yaml_snapshot!(examples);
    }

    #[test]
    fn keys_and_values() {
        let examples: Vec<_> = [
            "key: value",
            "x:y",
            "x :y",
            "x:5",
            "x:-4",
            "str:-4b",
            "nil:nil",
            "nil:",
            "still_nil:
             x: y",
            "
            key: value with : in it
            key_2: value with: in it
            ",
            "keys: {
              in:
              objects: []
            }"
        ]
        .into_iter()
        .map(|example| example.to_string() + "\n")
        .map(|str| parse_metadata(&str))
        .try_collect()
        .unwrap();

        assert_yaml_snapshot!(examples);
    }
}
