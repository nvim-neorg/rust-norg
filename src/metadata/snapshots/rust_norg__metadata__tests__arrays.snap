---
source: src/metadata/mod.rs
expression: examples
---
- Object:
    arr:
      Array: []
    empty_arr:
      Array: []
- Object:
    categories:
      Array:
        - Str: one
        - Str: two
        - Num: 45
- Object:
    arr:
      Array:
        - Str: arrays can contain everything
        - Num: 5
        - Num: -5
        - Num: 6020000000000000000000000000
        - Nil
        - Object:
            a:
              Array:
                - Str: b
            x:
              Str: y
        - Array: []
        - Array:
            - Str: hi
            - Str: hi
- Object:
    a2:
      Array: []
    a3:
      Array:
        - Str: hi
    arr:
      Array: []
    x:
      Str: y
